apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -576,29 +608,96 @@ async Task PrepareNewSessionAsync()
                         minimumSizeWaiters.ForEach(tcs => tcs.TrySetException(e));
                     }
                 }
+            } 
+
+            private Task GetNursePoolBackToHealthTask(CancellationToken cancellationToken)
```

Comment:
I'm slightly nervous about this, just because it's a big new chunk of code. Was the previous code simpler because we could always just create a single new session?

Reply:
Actually I think the previous code would have benefited from something like this as well. When the pool was unhealthy, we attempted to create a single new session for each caller trying to acquire a session. If these were "a lot" and depending on the options setup, we could have ended up either with more than the allowed number of sessions, or potentially "choking" with individual new session requests an already unhealthy pool/service. When creating sessions in batch, the chances of ending up with more sessions than allowed is greater, that's why I went this way. 
This method is really not that different from the `StartSessionCreationTasksIfNecessary.PrepareNewSessionsAsync` method, it does the same thing in different circumstances.
Does that sound good?

Reply:
Okay, that sounds fine.